samples/matter/common/src/watchdog/watchdog.cpp

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "watchdog/watchdog.h"
+
+#include <platform/CHIPDeviceLayer.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_CHIP_APP_LOG_LEVEL);
+
+using namespace Nrf::Watchdog;
+
+namespace
+{
+const struct device *const wdt = DEVICE_DT_GET(DT_ALIAS(watchdog0));
+bool sEnabled = false;
+struct WatchdogListItem : public sys_snode_t {
+	WatchdogSource *mWatchdog = nullptr;
+};
+sys_slist_t sWatchdogList;
+} // namespace
+
+WatchdogSource::WatchdogSource(uint32_t feedingInterval, FeedingCallback callback)
+	: mFeedingTime(feedingInterval), mCallback(callback)
+{
+	/* Check whether the watchdog peripheral is enabled and ready */
+	assert(device_is_ready(wdt));
+
+	/* Stop the Global Watchdog before further operations */
+	bool startAutomatically = false;
+	if (sEnabled) {
+		assert(Disable());
+		startAutomatically = true;
+	}
+
+	wdt_timeout_cfg wdtConfig = {};
+	/* Reset SoC when watchdog timer expires. */
+	wdtConfig.flags = WDT_FLAG_RESET_SOC;
+	/* The minimum window time must be 0 */
+	wdtConfig.window.min = 0;
+	/* The maximum window time must be equal for all channels */
+	wdtConfig.window.max = CONFIG_NCS_SAMPLE_MATTER_WATCHDOG_TIMEOUT;
+	/* The callback has no sense, because watchdog interrupt takes only ~64 us and CPU
+	 * cannot do anything useful */
+	wdtConfig.callback = nullptr;
+
+	WatchdogListItem *item = chip::Platform::New<WatchdogListItem>();
+	assert(item);
+
+	mChannel = wdt_install_timeout(wdt, &wdtConfig);
+	if (-ENOMEM == mChannel) {
+		LOG_ERR("Cannot create a new Watchdog source because there is no more available Watchdog channels");
```

Comment:
Should we exit then? We shouldn't rely on `asserts` for handling error cases that may occur even in a correctly written software. Asserts can be disabled. Also, it's better to use Zephyr `__ASSERTs` than C ones to have a better control on their behavior (but only for code internal assumption violations, not "normal" things like heap allocation failures).

Reply:
I've extended those assertions, but I'm still considering whether it is a good approach at all to make it all in a constructor. It works fine, and it is super simple for the user, but I'm wondering whether the errors will be noticed by developers.